# Lab 4, Exercise 1

In [1]:
import numpy as np
import pandas as pd

## Load data 

Here is an example of the first couple rows from the data:

| id | dur       | proto | service | state | spkts | dpkts | sbytes | dbytes | rate        | sttl | dttl | sload     | dload | sloss | dloss | sinpkt | dinpkt | sjit | djit | swin | stcpb | dtcpb | dwin | tcprtt | synack | ackdat | smean | dmean | trans\_depth | response\_body\_len | ct\_srv\_src | ct\_state\_ttl | ct\_dst\_ltm | ct\_src\_dport\_ltm | ct\_dst\_sport\_ltm | ct\_dst\_src\_ltm | is\_ftp\_login | ct\_ftp\_cmd | ct\_flw\_http\_mthd | ct\_src\_ltm | ct\_srv\_dst | is\_sm\_ips\_ports | attack\_cat | label |
|----|-----------|-------|---------|-------|-------|-------|--------|--------|-------------|------|------|-----------|-------|-------|-------|--------|--------|------|------|------|-------|-------|------|--------|--------|--------|-------|-------|--------------|---------------------|--------------|----------------|--------------|---------------------|---------------------|-------------------|----------------|--------------|---------------------|--------------|--------------|--------------------|-------------|-------|
| 1  | 0\.000011 | udp   | \-      | INT   | 2     | 0     | 496    | 0      | 90909\.0902 | 254  | 0    | 180363632 | 0     | 0     | 0     | 0\.011 | 0      | 0    | 0    | 0    | 0     | 0     | 0    | 0      | 0      | 0      | 248   | 0     | 0            | 0                   | 2            | 2              | 1            | 1                   | 1                   | 2                 | 0              | 0            | 0                   | 1            | 2            | 0                  | Normal      | 0     |


In [2]:
# Load the training and test data corresponding to exercise 1
# Create two separate pandas dataframes for the training and test data
# For each dataframe, import the following CSV data
# training set: 'data/exercise1/UNSW_NB15_training-set.csv'
# test set: 'data/exercise1/UNSW_NB15_testing-set.csv'

# CODE HERE

train_df = pd.read_csv("data/exercise1/UNSW_NB15_training-set.csv")
test_df = pd.read_csv("data/exercise1/UNSW_NB15_testing-set.csv")

train_df


id       dur proto service state  spkts  dpkts  sbytes  dbytes  \
0          1  0.000011   udp       -   INT      2      0     496       0   
1          2  0.000008   udp       -   INT      2      0    1762       0   
2          3  0.000005   udp       -   INT      2      0    1068       0   
3          4  0.000006   udp       -   INT      2      0     900       0   
4          5  0.000010   udp       -   INT      2      0    2126       0   
...      ...       ...   ...     ...   ...    ...    ...     ...     ...   
82327  82328  0.000005   udp       -   INT      2      0     104       0   
82328  82329  1.106101   tcp       -   FIN     20      8   18062     354   
82329  82330  0.000000   arp       -   INT      1      0      46       0   
82330  82331  0.000000   arp       -   INT      1      0      46       0   
82331  82332  0.000009   udp       -   INT      2      0     104       0   

                rate  ...  ct_dst_sport_ltm  ct_dst_src_ltm  is_ftp_login  \
0       90909.090200  ...                 1               2             0   
1      125000.000300  ...                 1               2             0   
2      200000.005100  ...                 1               3             0   
3      166666.660800  ...                 1               3             0   
4      100000.002500  ...                 1               3             0   
...              ...  ...               ...             ...           ...   
82327  200000.005100  ...                 1               2             0   
82328      24.410067  ...                 1               1             0   
82329       0.000000  ...                 1               1             0   
82330       0.000000  ...                 1               1             0   
82331  111111.107200  ...                 1               1             0   

       ct_ftp_cmd  ct_flw_http_mthd  ct_src_ltm  ct_srv_dst  is_sm_ips_ports  \
0               0                 0           1           2                0   
1               0                 0           1           2                0   
2               0                 0           1           3                0   
3               0                 0           2           3                0   
4               0                 0           2           3                0   
...           ...               ...         ...         ...              ...   
82327           0                 0           2           1                0   
82328           0                 0           3           2                0   
82329           0                 0           1           1                1   
82330           0                 0           1           1                1   
82331           0                 0           1           1                0   

       attack_cat  label  
0          Normal      0  
1          Normal      0  
2          Normal      0  
3          Normal      0  
4          Normal      0  
...           ...    ...  
82327      Normal      0  
82328      Normal      0  
82329      Normal      0  
82330      Normal      0  
82331      Normal      0  

[82332 rows x 45 columns]

### Part 1 
Keep the two dataframes separate and create train/test data and labels.  This will be used to experiment with the case where there are different types of activities in the training versus test set.

In [8]:
# Do all the following operations in the cell on both the dataframes separately
# 1) Keep only the datapoints where the 'attack_cat' column is equal to either 'Normal' or 'Fuzzers'
# CODE HERE

train_df = train_df.loc[train_df["attack_cat"].isin(['Normal','Fuzzers'])]
test_df = test_df.loc[test_df["attack_cat"].isin(['Normal','Fuzzers'])]

# 2) Get the labels from the dataframe (i.e., the values in the 'attack_cat' column)
# CODE HERE

train_labels = train_df["attack_cat"]
test_labels = test_df["attack_cat"]

# 3) Keep only the features we care about for this experiment.
# We only care about the numerical features between column 'spkts' and 'is_sm_ips_ports' (inclusive)
# CODE HERE

train_data = train_df.loc[:, "spkts":"is_sm_ips_ports"]
test_data = test_df.loc[:, "spkts":"is_sm_ips_ports"]

# You should now have four inputs usable for scikit-learn:
# training data
# training labels
# test data
# test labels
# Hint: You may have to add some minor code in the above to get the data ready for scikit-learn

### Part 2 
Create a new training/test split by combining the dataframes into one.  Then split the dataframe randomly into train/test data and labels.  This will be used to experiment with the case where there are largely the same types of activities in the training and test set.

In [4]:
# Combine the dataframes into a single dataframe, then do the following
# CODE HERE

combined_df = pd.concat([train_df, test_df], ignore_index=True)

# 1) Keep only the datapoints where the 'attack_cat' column is equal to either 'Normal' or 'Fuzzers'
# CODE HERE

combined_df = combined_df.loc[combined_df["attack_cat"].isin(['Normal','Fuzzers'])]

# 2) Get the labels from the dataframe (i.e., the values in the 'attack_cat' column)
# CODE HERE

combined_labels = combined_df["attack_cat"]

# 3) Keep only features we care about for this experiment.
# We only care about the numerical features between column 'spkts' and 'is_sm_ips_ports' (inclusive)
# CODE HERE

combined_data = combined_df.loc[:, "spkts":"is_sm_ips_ports"]

# 4) Create a random split; put 50% of the data into the training set and the other 50% into the test set
# Use scikit-learn's 'train_test_split'
# Hint: You may have to add some minor code in the above to get the data ready for scikit-learn's 'train_test_split'
# CODE HERE

from sklearn.model_selection import train_test_split

rand_train_data, rand_test_data, rand_train_labels, rand_test_labels = train_test_split(combined_data, combined_labels, test_size=0.5);

# You should now have four inputs usable for scikit-learn that are distinct from the inputs you created in Part 1.
# The inputs should correspond to:
# training data
# training labels
# test data
# test labels

## Train a classifier

In [5]:
# For each of the train/test splits, create a separate random forest with default sklearn parameters
# Hint: Ignoring import statements, each random forest can be created in a single line of code
# CODE HERE

from sklearn.ensemble import RandomForestClassifier

train_classifier = RandomForestClassifier().fit(train_data, train_labels)
rand_train_classifier = RandomForestClassifier().fit(rand_train_data, rand_train_labels)


/Users/bcheung/Documents/SoftwareProjects/UT/EE379K/lab4/venv/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/Users/bcheung/Documents/SoftwareProjects/UT/EE379K/lab4/venv/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


## Inference and results

In [6]:
# For each of the train/test splits and associated random forest, do the following:

# 1) Predict labels on the training data
# CODE HERE

pred_labels = train_classifier.predict(train_data)
rand_pred_labels = rand_train_classifier.predict(rand_train_data)

# 2) Print metrics on the training data; use sklearn's implementation of precision, recall, f1, and accuracy
# CODE HERE

from sklearn import metrics

def print_metrics(title, test_labels, pred_labels):
    
    precision = metrics.precision_score(test_labels, pred_labels, pos_label="Fuzzers")
    recall = metrics.recall_score(test_labels, pred_labels, pos_label="Fuzzers")
    f1measure = metrics.f1_score(test_labels, pred_labels, pos_label="Fuzzers")
    accuracy = metrics.accuracy_score(test_labels, pred_labels)

    print(title)
    print(' precision = ' + str(precision)) # true positives / predicted positives
    print('    recall = ' + str(recall))    # true positives / actual positives
    print('F1-measure = ' + str(f1measure)) # weighted average of the precision and recall
    print('  accuracy = ' + str(accuracy))  # correctly pred / sample size
    print('\n')

print_metrics("train metrics", train_labels, pred_labels)
print_metrics("rand train metrics", rand_train_labels, rand_pred_labels)
    
# 3) Predict labels on the test data
# CODE HERE

test_pred_labels = train_classifier.predict(test_data)
test_rand_pred_labels = rand_train_classifier.predict(rand_test_data)

# 4) Print metrics on the test data; again, use precision, recall, f1, and accuracy
# CODE HERE

print_metrics("test metrics", test_labels, test_pred_labels)
print_metrics("rand test metrics", rand_test_labels, test_rand_pred_labels)


train metrics
 precision = 0.9960363336085879
    recall = 0.9948861761794787
F1-measure = 0.995460922670628
  accuracy = 0.9987227718173796


rand train metrics
 precision = 0.9766699410609038
    recall = 0.9819753086419754
F1-measure = 0.9793154395469097
  accuracy = 0.9914026917762653


test metrics
 precision = 0.8431095406360424
    recall = 0.26242850857897054
F1-measure = 0.40026841134037916
  accuracy = 0.8072360616844603


rand test metrics
 precision = 0.728255684528542
    recall = 0.7572751322751323
F1-measure = 0.7424819648212695
  accuracy = 0.8916125070364874




## Questions:

1) For results using Part 1 data, what is the precision and recall?

Results for the training data:
    Precision: 0.996
    Recall: 0.995
    
Results for the test data:
    Precision: 0.843
    Recall: 0.262

2) For results using Part 1 data, describe the difference in the results on the training and test data. What does this signify?

Predicting the labels on the training data was quite accurate. However, the predicting on the test data was less accurate. The model had a significantly lower recall score and F1-measure. These results may have been caused by overfitting, since the classifier was trained with only one set of data.

3) What changes in the results on the test data once you combine the data for Part 2? Does this produce a better classifier? Why or why not?

The recall score, F1-measure, and accuracy improve after combining and randomly spliting the data. This classifier is better overall because it is more accurate in predicating and identifying fuzzing traffic (accuracy and recall are higher) even though it has more false positives (precision is lower). It is better to over identify for fuzzing traffic than to under identify so that you can act accordingly to be safe.